### PR TITLE
Tests Generelization for multiple accelerator devices

### PR DESCRIPTION
**Pull request by rahulsingh-intel**
Motivation: Generalize unit tests so that can be executed for cuda and non cuda devices.
Chnages: There are general changes in common_dtesnor module for device type generalization so that tests can be executed on non cuda devices too.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

**Comment by rahulsingh-intel** [comment]
> @pytorchbot label "topic: not user facing"

@pytorchbot label "topic: not user facing"

**Comment by rahulsingh-intel** [comment]
> > @pytorchbot label "topic: not user facing"
> 
> @pytorchbot label "topic: not user facing"

hi @kwen2501 , @fegin please review the changes

**Comment by rahulsingh-intel** [comment]
> * Check Labels / Check labels (pull_request_target)### [Check Labels / Check labels (pull_request_target)](/pytorch/pytorch/actions/runs/11683434511/job/32532607434?pr=139749)

@pytorchbot  rebase

**Comment by rahulsingh-intel** [comment]
> Successfully rebased `dtensor_common` onto `refs/remotes/origin/viable/strict`, please pull locally before adding more changes (for example, via `git checkout dtensor_common && git pull --rebase`)

hi @kwen2501 , can you please review and approve the changes

File: test/distributed/_tensor/test_dtensor_compile.py [hunk 1]
```diff
@@ -870,9 +875,10 @@ def test_2d_fsdp_tp_compile(self):
             "mlp_1.net2": RowwiseParallel(),
         }
         tp_model = parallelize_module(model, twod_mesh["tp"], parallelize_plan)
+        device_id = "hpu:0" if TEST_HPU else self.rank
```
Comment (kwen2501) [inline test/distributed/_tensor/test_dtensor_compile.py:878]: Can you make `device_id` a property of the class?
```
@property
def device_id(self) -> torch.device:
    device_count = torch.get_device_module(self.device_type).device_count()
    return torch.device(device_type, self.rank % device_count)
```
Thanks!

File: test/distributed/_tensor/test_random_ops.py [hunk 2]
```diff
@@ -348,4 +349,4 @@ def test_meta_tensor_init(self):
 
 
 if __name__ == "__main__":
-    run_tests()
+    run_tests()
```
Comment (kwen2501) [inline test/distributed/_tensor/test_random_ops.py:352]: Same

File: test/distributed/_tensor/test_redistribute.py [hunk 3]
```diff
@@ -366,7 +366,7 @@ def test_redistribute_shard_dim_change(self):
                 local_out_dt = out_dt.to_local()
                 local_expected_dt = expected_dt.to_local()
                 self.assertEqual(out_dt.to_local(), expected_dt.to_local())
-                if self.device_type == "cuda":
+                if self.device_type == "cuda" or self.device_type == "hpu":
```
Comment (kwen2501) [inline test/distributed/_tensor/test_redistribute.py:369]: I wonder if we could truly generalize this line than adding another device string here?

File: torch/testing/_internal/distributed/_tensor/common_dtensor.py [hunk 4]
```diff
@@ -31,17 +35,27 @@
...
+DEVICE_COUNT = _get_device_module("cuda").device_count() if TEST_CUDA else _get_device_module("hpu").device_count()
 
 DEVICE_TYPE = (
-    "cuda" if torch.cuda.is_available() and torch.cuda.device_count() > 1 else "cpu"
+    "cuda"
+    if TEST_CUDA and DEVICE_COUNT > 1
+    else "hpu"
+    if TEST_HPU and DEVICE_COUNT > 1
+    else "cpu"
+)
```
Comment (kwen2501) [inline torch/testing/_internal/distributed/_tensor/common_dtensor.py:48]: I wonder if we could first define `DEVICE_TYPE` then `DEVICE_COUNT` so that we can call `_get_device_module(DEVICE_TYPE)` then `device_count()`?

File: torch/testing/_internal/distributed/_tensor/common_dtensor.py [hunk 5]
```diff
@@ -31,17 +35,27 @@
...
+PG_BACKEND = (
+    "nccl" if DEVICE_TYPE == "cuda" else "hccl" if DEVICE_TYPE == "hpu" else "gloo"
 )
```
Comment (kwen2501) [inline torch/testing/_internal/distributed/_tensor/common_dtensor.py:51]: I wonder if we could build a map instead? Thanks!

File: torch/testing/_internal/distributed/_tensor/common_dtensor.py [hunk 6]
```diff
@@ -301,17 +315,17 @@ def world_size(self) -> int:
 
     @property
     def backend(self) -> str:
-        backend = "nccl" if self.device_type == "cuda" else "gloo"
+        backend = "nccl" if self.device_type == "cuda" else "hccl" if self.device_type == "hpu" else "gloo"
```
Comment (kwen2501) [inline torch/testing/_internal/distributed/_tensor/common_dtensor.py:318]: How about reusing the `PG_BACKEND` map above?

**Review by kwen2501 — COMMENTED** [review]
Thanks for the effort. Overall looks good to me. I left a couple comments. I think the goal is to use `device_type` as much as possible and reduce the dependency on `cuda` or `hpu` strings. You did that well in most places, just need to cover the few ones left. Thanks!

File: test/distributed/_tensor/test_convolution_ops.py [hunk 7]
```diff
@@ -164,23 +159,25 @@ def test_depthwise_convolution(self):
         bias_mse_abs = torch.mean(bias_diff_abs * bias_diff_abs).item()
         weight_mse_rel = torch.mean(weight_diff_rel * weight_diff_rel).item()
         bias_mse_rel = torch.mean(bias_diff_rel * bias_diff_rel).item()
+        abs_value = 1e-4 if TEST_HPU else 1e-6
```
[the remark this answers is not in the document]
Comment (rahulsingh-intel) [reply test/distributed/_tensor/test_convolution_ops.py]: corrected.

File: torch/testing/_internal/distributed/_tensor/common_dtensor.py (hunk 6 above)
Comment (zhangxiaoli73) [reply torch/testing/_internal/distributed/_tensor/common_dtensor.py:318]: If we still keep those `if...else` statement, then a mapping list should be maintained and changed for each new backend added even it's out-of-tree.  

Then, is it possible to use `undefined` backend which will find corresponding device backend by device type in process group init? e.g. `nccl` for cuda, `gloo` for cpu, `hccl` for hpu and `xccl` for xpu if `hccl` and `xccl` is registered out-of-tree.

In code logic,
1. Init process group with `undefined` backend.
https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L1569 
2. Register available backend in `backend_config.get_device_backend_map()` to process group
https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L1818
3. `undefined` backend will use default device backend map , which is `gloo` for cpu, `nccl` for cuda. Out-of-tree backend will also be added to this default map by https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L254

Comment (ankurneog) [reply torch/testing/_internal/distributed/_tensor/common_dtensor.py]: > If we still keep those `if...else` statement, then a mapping list should be maintained and changed for each new backend added even it's out-of-tree.
> 
> Then, is it possible to use `undefined` backend which will find corresponding device backend by device type in process group init? e.g. `nccl` for cuda, `gloo` for cpu, `hccl` for hpu and `xccl` for xpu if `hccl` and `xccl` is registered out-of-tree.
> 
> In code logic,
> 
> 1. Init process group with `undefined` backend.
>    https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L1569
> 2. By backend name to query a `backend_config` which has a `device_backend_map`.
>    https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L1816
> 
> `undefined` backend name will use default device backend map , which is `gloo` for cpu, `nccl` for cuda. Out-of-tree backend will also be added to this default map by https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L254
> 
> 3. Register available backend in `backend_config.get_device_backend_map()` to process group
>    https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L1818

This can be straight one -to-one mapping for default backends : 
https://github.com/pytorch/pytorch/pull/140536

Comment (rahulsingh-intel) [reply torch/testing/_internal/distributed/_tensor/common_dtensor.py]: hi @kwen2501  please review.

Comment (rahulsingh-intel) [reply torch/testing/_internal/distributed/_tensor/common_dtensor.py]: hi @kwen2501 CI ran fine, please approve after review.

**Comment by rahulsingh-intel** [comment]
> Successfully rebased `dtensor_common` onto `refs/remotes/origin/viable/strict`, please pull locally before adding more changes (for example, via `git checkout dtensor_common && git pull --rebase`)

hi @kwen2501  please review.

**Comment by ankurneog** [comment]
@kwen2501 : Could you please help with the approval , it would be good if we can push this before the code freeze for v2.6.0. Thank you

**Comment by rahulsingh-intel** [comment]
@kwen2501 Gentle reminder !

**Comment by ankurneog** [comment]
@kwen2501  : Gentle reminder, could you please help with the approval. thank you.

**Review by kwen2501 — APPROVED** [review]
LGTM. Sorry about the delay

**Comment by c-p-i-o** [comment]
Oops. I think the diff is contributing to a test failure here?
https://github.com/pytorch/pytorch/actions/runs/12671761436/job/35348294544?pr=139749

**Comment by kwen2501** [comment]
If facing any lint error, you can use `lintrunner -a` to fix it.

**Comment by rahulsingh-intel** [comment]
> If facing any lint error, you can use `lintrunner -a` to fix it. Also, not sure how @pytorchmergebot diffs from @pytorchbot . We usually use: `@pytorchbot merge`

There was no lint error while uploading. Will use pytorchbot next time , Thanks Kwen.

**Comment by guangyey** [comment]
"Try to land this since the failure is unrelated."
@pytorchbot merge
